### DISCUSSION TITLE
I want to verify a signature created with ethers.js 6.7.1

### DISCUSSION THREAD

**Discussion by javaspeak** [web3j/web3j#1964] [Q&A] [answered]
On the front end I am using react, Wallet Connect 2 and ethers.js 6.7.1

Currently am creating a signature on the front end using ethers.js 6.7.1.  I am then retrieving the wallet address from the signature to verify it is the same address I signed with.   I wish to move the verifying bit of the signature code to the Java Backend so was thinking of using the web3j library for that.  This question is about how do I verify the signature using web3j? 

I am showing the JavaScript code used to create the signature so that the verify signature java code can be compatible.

My TypeScript code which creates the signature and verifies it is as follows:

```
const connectWallet = async () => {
    await open();
    const unWatch = ethereumClient.watchAccount(
      (account) => {
        if (account.isConnected) {
          let address = account.address;

          if (address) {
            setWalletAddress(address);
          }

          unWatch();

          account.connector?.getProvider().then(
            (provider) => {
              const browserProvider = new BrowserProvider(provider);

              browserProvider.getSigner().then(
                (signer: JsonRpcSigner) => {

                  if (secretForSigning) {
                    signer.signMessage(secretForSigning).then(
                      (signedMessage) => {
                        setSignature(signedMessage);
                        const signerAddress = ethers.verifyMessage(secretForSigning, signedMessage);

                        if (address === signerAddress) {
                          console.log('Verified')
                        }
                        else {
                          console.log('addresses different');
                          console.log('signerAddress', signerAddress);
                        }
                      }
                    );
                  }
                }
              );
            }
          );
        }
      }
    );
  }
```

My code successfully creates the signature using ethers.js.  Although the code is client side I also verify the signature to check the ethers code is working:

```
const signerAddress = ethers.verifyMessage(secretForSigning, signedMessage);

if (address === signerAddress) {
    console.log('Verified')
}
```

What I want to do is verify the signature on the Java Side instead of on the TypeScript side.  As you can see I can verify the signature and get the wallet address with a single line:

`const signerAddress = ethers.verifyMessage(secretForSigning, signedMessage);`

My question is how do I do the same thing using web3j ?

Thank you for your time.

**Comment by javaspeak** [comment]
I am not sure if I need to go low level and be adding stuff like this to the signedMessage before trying to process it:

```
String GETH_SIGN_PREFIX = "\u0019Ethereum Signed Message:\n32";
byte[] msgBytes = new byte[GETH_SIGN_PREFIX.getBytes().length + messageBytes.length];
byte[] prefixBytes = GETH_SIGN_PREFIX.getBytes();
System.arraycopy( prefixBytes, 0, msgBytes, 0, prefixBytes.length );
System.arraycopy( messageBytes, 0, msgBytes, prefixBytes.length, messageHashBytes.length );
```

**Comment by NickSneo** [comment, marked as answer]
Hey @javaspeak , Have you tried these functions to recover or verify - 
`signedMessageToKey` , `recoverFromSignature` ,  `getRecId`, etc from Sign Class - https://github.com/web3j/web3j/blob/master/crypto/src/main/java/org/web3j/crypto/Sign.java
